substrate/client/rpc-spec-v2/src/chain_head/chain_head_storage.rs

```diff
@@ -18,192 +18,106 @@
-	operation_max_storage_items: usize,
 	_phandom: PhantomData<(BE, Block)>,
 }
 
+impl<Client, Block, BE> Clone for ChainHeadStorage<Client, Block, BE> {
+	fn clone(&self) -> Self {
+		Self { client: self.client.clone(), _phandom: PhantomData }
+	}
+}
+
 impl<Client, Block, BE> ChainHeadStorage<Client, Block, BE> {
 	/// Constructs a new [`ChainHeadStorage`].
-	pub fn new(client: Arc<Client>, operation_max_storage_items: usize) -> Self {
-		Self {
-			client: Storage::new(client),
-			iter_operations: VecDeque::new(),
-			operation_max_storage_items,
-			_phandom: PhantomData,
-		}
+	pub fn new(client: Arc<Client>) -> Self {
+		Self { client: Storage::new(client), _phandom: PhantomData }
 	}
 }
 
 impl<Client, Block, BE> ChainHeadStorage<Client, Block, BE>
 where
-	Block: BlockT + 'static,
-	BE: Backend<Block> + 'static,
-	Client: StorageProvider<Block, BE> + 'static,
+	Block: BlockT + Send + 'static,
+	BE: Backend<Block> + Send + 'static,
+	Client: StorageProvider<Block, BE> + Send + Sync + 'static,
 {
-	/// Iterate over (key, hash) and (key, value) generating the `WaitingForContinue` event if
-	/// necessary.
-	async fn generate_storage_iter_events(
-		&mut self,
-		mut block_guard: BlockGuard<Block, BE>,
-		hash: Block::Hash,
-		child_key: Option<ChildInfo>,
-	) {
-		let sender = block_guard.response_sender();
-		let operation = block_guard.operation();
-
-		while let Some(query) = self.iter_operations.pop_front() {
-			if operation.was_stopped() {
-				return
-			}
-
-			let result = self.client.query_iter_pagination(
-				query,
-				hash,
-				child_key.as_ref(),
-				self.operation_max_storage_items,
-			);
-			let (events, maybe_next_query) = match result {
-				QueryIterResult::Ok(result) => result,
-				QueryIterResult::Err(error) => {
-					send_error::<Block>(&sender, operation.operation_id(), error.to_string());
-					return
-				},
-			};
-
-			if !events.is_empty() {
-				// Send back the results of the iteration produced so far.
-				let _ = sender.unbounded_send(FollowEvent::<Block::Hash>::OperationStorageItems(
-					OperationStorageItems { operation_id: operation.operation_id(), items: events },
-				));
-			}
-
-			if let Some(next_query) = maybe_next_query {
-				let _ =
-					sender.unbounded_send(FollowEvent::<Block::Hash>::OperationWaitingForContinue(
-						OperationId { operation_id: operation.operation_id() },
-					));
-
-				// The operation might be continued or cancelled only after the
-				// `OperationWaitingForContinue` is generated above.
-				operation.wait_for_continue().await;
-
-				// Give a chance for the other items to advance next time.
-				self.iter_operations.push_back(next_query);
-			}
-		}
-
-		if operation.was_stopped() {
-			return
-		}
-
-		let _ =
-			sender.unbounded_send(FollowEvent::<Block::Hash>::OperationStorageDone(OperationId {
-				operation_id: operation.operation_id(),
-			}));
-	}
-
 	/// Generate the block events for the `chainHead_storage` method.
 	pub async fn generate_events(
```

Comment:
I wasn't happy with this generate_events and generate_events_iter, so now I have merged them which provides proper backpressure all storage queries